core/src/test/java/org/elasticsearch/action/support/replication/TransportWriteActionTests.java

```diff
@@ -124,15 +126,46 @@ public void testReplicaWaitForRefresh() throws Exception {
         boolean forcedRefresh = randomBoolean();
         refreshListener.getValue().accept(forcedRefresh);
         assertNotNull(listener.response);
+        assertNull(listener.failure);
         resultChecker.accept(listener.response, forcedRefresh);
     }
 
+    public void testDocumentFailureInShardOperationOnPrimary() throws Exception {
+        handleDocumentFailure(new TestAction(true, true), TestAction::shardOperationOnPrimary, TestAction.WritePrimaryResult::respond);
+    }
+
+    public void testDocumentFailureInShardOperationOnReplica() throws Exception {
+        handleDocumentFailure(new TestAction(randomBoolean(), true), TestAction::shardOperationOnReplica,
+                TestAction.WriteReplicaResult::respond);
+    }
+
+    private <Result, Response> void handleDocumentFailure(TestAction testAction,
```

Comment:
nit: can you unpack it? it saves on 10 lines of code at the expense of making it much harder to understand what it does - you end going back and forth - I think it's not worth it?